File: 144_Binary_tree_preorder_traversal/solution_recursion.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode *root) {
        //recursive
        vector<int> v,temp;
        if(root==NULL){
            return v;
        }else{
            v.push_back(root->val);
        }
        temp = preorderTraversal(root->left);
        v.insert(v.end(),temp.begin(),temp.end());
        temp = preorderTraversal(root->right);
        v.insert(v.end(),temp.begin(),temp.end());
        return v;
    }
};

File: 144_Binary_tree_preorder_traversal/solution_recursion.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def preorderTraversal(self, root):
        #recursion
        l = []
        if root is None: return l
        l.append(root.val)
        l.extend(self.preorderTraversal(root.left))
        l.extend(self.preorderTraversal(root.right))
        return l
        